### PR TITLE
Catch DottedVariable TypeError

### PR DESCRIPTION
**Pull request by EmilyBourne**
Raise an error when DottedVariable receives unexpected types. This is needed to handle the following:
```
a = " ".join(["boy","girl"])
```
See issue #605 

Raise an error when a class method is called on an object which has no class methods known to pyccel. This is needed to handle the following:
```
a = " "
b = a.join(["boy","girl"])
```

### CODE REVIEW

**Review by saidctb — COMMENTED** [review]
instead of try except clause, can yo just do a check if the attribut exist ?, either in the DottedVariable contructor or in the _visit_attribute

**Comment by EmilyBourne** [comment]
> instead of try except clause, can you just do a check if the attribute exist ?, either in the DottedVariable constructor or in the _visit_attribute, also shouldn't this be a sematics error and not a syntacitic one 

The error is caused at the syntactic stage by trying to make the lhs of a DottedVariable be a LiteralString. We probably get the same error if we do something like `1j.imag`. Either we need to test for the types accepted by DottedVariable (but that is just duplicating the type check so a try/except is more maintainable). Or we need to accept literals as the lhs of a DottedVariable and raise a similar error in the semantic step.

This seemed like the simplest option for now

**Comment by saidctb** [comment]
> > instead of try except clause, can you just do a check if the attribute exist ?, either in the DottedVariable constructor or in the _visit_attribute, also shouldn't this be a sematics error and not a syntacitic one
> 
> The error is caused at the syntactic stage by trying to make the lhs of a DottedVariable be a LiteralString. We probably get the same error if we do something like `1j.imag`. Either we need to test for the types accepted by DottedVariable (but that is just duplicating the type check so a try/except is more maintainable). Or we need to accept literals as the lhs of a DottedVariable and raise a similar error in the semantic step.
> 
> This seemed like the simplest option for now
I think a DottedVariable should accept any literal, so the error will no longer be there, and raise the error in the sematics stage, this seems simple to me , I'm not even sure that we should check anything in the syntacitc stage the DottedVariable should accept anything

**Comment by saidctb** [comment]
can you also treat this case ?
```python
a = "  "
b = a.join(["boy","girl"])
```

File: pyccel/parser/semantic.py [hunk 1]
```diff
@@ -1006,10 +1006,17 @@ def _visit_Symbol(self, expr, **settings):
 
     def _visit_DottedVariable(self, expr, **settings):
 
+        if isinstance(expr.lhs, Literal):
```
Comment (saidctb) [inline pyccel/parser/semantic.py:1009]: probably , you don't need this anymore. after adding the second error report

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1009]: I still need this as a `Literal` currently doesn't have a `cls_base`

Comment (saidctb) [reply pyccel/parser/semantic.py:1009]: can you do this `if hasattr(first, 'cls_base') and  first.cls_base is None:`, with this you can have both cases

File: pyccel/parser/semantic.py [hunk 2]
```diff
@@ -1060,6 +1060,12 @@ def _visit_DottedVariable(self, expr, **settings):
 
         # look for a class method
         if isinstance(expr.rhs, Application):
+            if not hasattr(first, 'cls_base') or first.cls_base is None:
+                errors.report('Class properties of object of type {} are not known'.format(first.dtype),
```
Comment (saidctb) [inline pyccel/parser/semantic.py:1064]: can you just raise and attribute error like python  `object has no attribute`

**Review by saidctb — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**